text/0039-local-keys.md

```diff
@@ -0,0 +1,153 @@
+
+Newly created will have a type set to "unit" to hide the implementation details and discourage using "read" or "write" APIs to operate on local keyspace. The host will also secure this possibility on the backend, and inside of these host functions, a particular check will verify and forbid direct operations on the `Key::Local` space.
+
+### read_local
+
+```rust
+fn read_local(uref, key) -> T;
+```
+
+Reads data from a local key under "URef" and "key" bytes and returns a stored value. The system then verifies that URef is valid in the caller context and performs hashing of the "URef" address and the key bytes.
+
+A secondary key index is not involved in this operation.
+
+### write_local
+
+```rust
+fn write_local(uref, key, data);
+```
+
+Writes data under a key to a local key referenced by "URef.". The system then verifies that URef is in the execution context, and an additional check is in place to confirm if the "WRITE" permission bit is on.
+
+If the given "key" is not present in the secondary index that "URef" points to, then it is extended with "key." bytes.
+
+### add_local
+
+```rust
+fn add_local(uref, key, data);
+```
+
+Similar to `write_local,` but emits an `ADD` transformation for a given local key.
+
+### list_local
+
+```rust
+fn list_local(uref) -> Vec<Vec<u8>>;
+```
+
+This function lists all keys under a Key::Local referenced by "URef." Internally it reads all the keys from the secondary index and returns them to the caller.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+As explained above current limitation of key lengths inside a trie requires us to maintain a secondary index of keys under a URef to retain the ability to migrate into the unhashed, more extended keyspace for more efficient iteration.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Alternatives to using proposed local keys would be to store data as separate named keys. Doing this creates another problem - mainly bloat of account/contract's named keys which may increase WASM execution startup times as account (or contract) has to be deserialized before execution.
+
+## Prior art
+
+[prior-art]: #prior-art
+
+In the past, we used to have `Key::Local,` which we first deprecated and then removed as we used it only in system contracts, and there was a better functionality in place to replace uses. This proposal builds on top of past implementation by fixing problems that it had along the way.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How do we name this feature to reflect better what it does? 
+- Does it make sense to make the provisioned URef a unit? Can we use it to store additional information related to the stored data under the local key?
```

Comment:
How does the migration of local storage look like? How will the upgraded contract access the pre-upgraded local data?

Reply:
This data migration will be transparent to already deployed contracts. I see this as an internal improvement, but the API would be forward-compatible.

During migration, we will iterate over reachable URefs (i.e., those stored under account or contract's named keys) detect if there's extra payload under that Unit with all the key bytes, then reindex the keys from the old hashed address (`hash(uref_addr || key_bytes)`), into `uref_addr || hash(key_bytes)` and remove the extra payload.

Reply:
OK, thanks. I just wanted to make sure this is not being overlooked and we already have a plan for how to solve it. We can't deliver local keys without also allowing contract owners to transfer them between contract upgrades.